statsmodels/tsa/base/tsa_model.py

```diff
@@ -319,9 +341,23 @@ def _get_index_loc(self, key, base_index=None):
                                             periods=len(index) + 1,
                                             freq=base_index.freq)
 
-        # Get the location (note that get_loc will throw a KeyError if key is
-        # invalid)
-        loc = index.get_loc(key)
+        # Get the location
+        if date_index:
+            # (note that get_loc will throw a KeyError if key is invalid)
+            loc = index.get_loc(key)
+        elif int_index or range_index:
+            # For Int64Index and RangeIndex, key is assumed to be the location
+            # and not an index value (this assumption is required to support
+            # RangeIndex)
+            try:
+                index[key]
+            # We want to raise a KeyError in this case, to keep the exception
+            # consistent across index types
+            except IndexError as e:
+                raise KeyError(str(e))
+            loc = key
+        else:
+            loc = index.get_loc(key)
```

Comment:
Out of scope, but commenting on 331/367 below: why is copying index necessary?  `pd.Index` objects are immutable(ish).

Reply:
I didn't realize that they were immutable. If so, I guess there's no reason we couldn't return the original index.